upstairs/src/client.rs

```diff
@@ -2332,165 +2333,324 @@ pub(crate) enum ClientRunResult {
     QueueClosed,
 }
 
-#[allow(clippy::too_many_arguments)]
-async fn client_run(
+/// Data structure to hold context for the client IO task
+///
+/// Client IO is managed by two tasks:
+/// - The tx task, which calls `ClientIoTask::run`, sends messages from the main
+///   task to the downstairs via a socket
+/// - The rx task, which is spawned within `ClientIoTask::run` (and is not
+///   publicly visible) receives messages from the socket and sends them
+///   directly to the main task.
+///
+/// Splitting tx and rx is important, because it means that one or the other
+/// should always be able to make progress.
+struct ClientIoTask {
     client_id: ClientId,
     tls_context: Option<Arc<crucible_common::x509::TLSContext>>,
     target: SocketAddr,
-    mut rx: mpsc::Receiver<ClientRequest>,
-    mut start: oneshot::Receiver<()>,
-    mut stop: oneshot::Receiver<ClientStopReason>,
+
+    /// Request channel from the main task
+    rx: mpsc::Receiver<ClientRequest>,
+
+    /// Reply channel to the main task
     tx: mpsc::Sender<ClientResponse>,
+
+    /// Oneshot used to start the task
+    start: oneshot::Receiver<()>,
+
+    /// Oneshot used to stop the task
+    stop: oneshot::Receiver<ClientStopReason>,
+
+    /// Delay on startup, to avoid a busy-loop if connections always fail
     delay: bool,
+
+    /// Handle for the rx task
+    recv_task: ClientRxTask,
+
     log: Logger,
-) {
-    let r = client_run_inner(
-        client_id,
-        tls_context,
-        target,
-        &mut rx,
-        &mut start,
-        &mut stop,
-        &tx,
-        delay,
-        &log,
-    )
-    .await;
-
-    warn!(log, "client task is sending Done({r:?})");
-    if tx.send(ClientResponse::Done(r)).await.is_err() {
-        warn!(
-            log,
-            "client task could not reply to main task; shutting down?"
-        );
+}
+
+/// Handle for the rx side of client IO
+///
+/// This is a convenient wrapper so that we can join the task exactly once,
+/// aborting if the wrapper is dropped without being joined.
+struct ClientRxTask {
+    handle: Option<tokio::task::JoinHandle<ClientRunResult>>,
+}
+
+impl ClientRxTask {
+    /// Waits for the client IO task to end
+    ///
+    /// # Panics
+    /// If the `JoinHandle` returns a `JoinError`, or this is called without an
+    /// IO handle (i.e. before the task is started or after it has been joined).
+    async fn join(&mut self) -> ClientRunResult {
+        match self.handle.as_mut() {
+            Some(t) => {
+                let r = t.await.expect("join error on recv_task");
```

Comment:
Maybe?  I'm not sure how that panic could ever be hit, even in our existing code: we only ever call `recv_task.abort()` in a branch that immediately exits (and doesn't return to the main `select!` which checks for task completion).

I'm wondering if this happens during program exit, where all bets are off...

Reply:
Quoth [the Tokio docs](https://docs.rs/tokio/latest/tokio/task/index.html#cancellation):
> Additionally, shutting down a Tokio runtime (e.g. by returning from #[tokio::main]) immediately cancels all tasks on it.

I think that the panic is because shutdown manages to cancel the `recv_task` while we're still awaiting it here.  I'm going to allow cancellation, but still panic on task panics here.

f118da1f905f1db2db506530d096f55723650183 (panic changes) + f6d96690dd474f071935a8e9dbcc4131398234b5 (for logging)

Reply:
You are correct in your response, I linked the wrong bug!

https://github.com/oxidecomputer/crucible/issues/1120 Is the one I wanted to link.